src/compute-engine/global-types.ts

```diff
@@ -1038,12 +1051,24 @@ export interface BoxedExpression {
    *
    * Infer the type of this expression.
    *
-   * If the type of this expression is already known, return `false`.
+   * Effective only for overall BoxedExpression types which are *non-constant* and therefore for
+   * which its value, and thereby type, can potentially vary.
+   *
+   * For symbols, inference may take place only for undeclared, or previously inferred symbols. For
```

Comment:
True, but with one caveat: a symbol declared without a type, or with a type `unknown` will still be subject to type inferrence.